## Ejercicio 1:

Descargar el dataset "Airlines Delay: Airline on-time statistics and delay causes". Cargarlo en un dataframe de Pandas, explicar los datos que contiene, las variables que hay, y quedarnos solo con las columnas relevantes, justificando el porqué. 

Nota importante: reducir la dimensión del dataset de forma aleatoria a solo 200000 registros y trabajar con este dataset reducido.

In [2]:
import pandas as pd
import numpy as np  # Lo usaremos para asegurarnos que no quedan "strings vacíos" que no sean reconocidos como "NaN".
# El motivo de lo anterior es que eliminaremos las filas que contengan NaN, pero los "strings vacíos" no se reconocen como tal.
import gc  # "Garbage collector", para liberar memoria luego de obtener la muestra de 200000 registros.

In [3]:
retrasos_full = pd.read_csv('DelayedFlights.csv')

In [4]:
retrasos_full.head()

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0           0  2008      1           3          4   2003.0        1955   
1           1  2008      1           3          4    754.0         735   
2           2  2008      1           3          4    628.0         620   
3           4  2008      1           3          4   1829.0        1755   
4           5  2008      1           3          4   1940.0        1915   

   ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
0   2211.0        2225            WN  ...     4.0     8.0          0   
1   1002.0        1000            WN  ...     5.0    10.0          0   
2    804.0         750            WN  ...     3.0    17.0          0   
3   1959.0        1925            WN  ...     3.0    10.0          0   
4   2121.0        2110            WN  ...     4.0    10.0          0   

   CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
0                 N         0           NaN           NaN      NaN   
1                 N         0           NaN           NaN      NaN   
2                 N         0           NaN           NaN      NaN   
3                 N         0           2.0           0.0      0.0   
4                 N         0           NaN           NaN      NaN   

  SecurityDelay  LateAircraftDelay  
0           NaN                NaN  
1           NaN                NaN  
2           NaN                NaN  
3           0.0               32.0  
4           NaN                NaN  

[5 rows x 30 columns]

In [5]:
retrasos_full.tail()

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
1936753     7009710  2008     12          13          6   1250.0        1220   
1936754     7009717  2008     12          13          6    657.0         600   
1936755     7009718  2008     12          13          6   1007.0         847   
1936756     7009726  2008     12          13          6   1251.0        1240   
1936757     7009727  2008     12          13          6   1110.0        1103   

         ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
1936753   1617.0        1552            DL  ...     9.0    18.0          0   
1936754    904.0         749            DL  ...    15.0    34.0          0   
1936755   1149.0        1010            DL  ...     8.0    32.0          0   
1936756   1446.0        1437            DL  ...    13.0    13.0          0   
1936757   1413.0        1418            DL  ...     8.0    11.0          0   

         CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
1936753                 N         0           3.0           0.0      0.0   
1936754                 N         0           0.0          57.0     18.0   
1936755                 N         0           1.0           0.0     19.0   
1936756                 N         0           NaN           NaN      NaN   
1936757                 N         0           NaN           NaN      NaN   

        SecurityDelay  LateAircraftDelay  
1936753           0.0               22.0  
1936754           0.0                0.0  
1936755           0.0               79.0  
1936756           NaN                NaN  
1936757           NaN                NaN  

[5 rows x 30 columns]

In [6]:
retrasos_full.columns.values

array(['Unnamed: 0', 'Year', 'Month', 'DayofMonth', 'DayOfWeek',
       'DepTime', 'CRSDepTime', 'ArrTime', 'CRSArrTime', 'UniqueCarrier',
       'FlightNum', 'TailNum', 'ActualElapsedTime', 'CRSElapsedTime',
       'AirTime', 'ArrDelay', 'DepDelay', 'Origin', 'Dest', 'Distance',
       'TaxiIn', 'TaxiOut', 'Cancelled', 'CancellationCode', 'Diverted',
       'CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay',
       'LateAircraftDelay'], dtype=object)

No muestra todas las columnas. Para analizar mejor los datos, haremos que las muestre todas.

In [7]:
len(retrasos_full.columns.values)  # Vemos cuántas columnas hay.

30

In [8]:
pd.set_option('display.max_columns', 30)  # Indicamos a Pandas que muestre las treinta.

In [9]:
# Revisamos el tipo de variable, ahora sí, viendo todas.
retrasos_full.head(10)

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0           0  2008      1           3          4   2003.0        1955   
1           1  2008      1           3          4    754.0         735   
2           2  2008      1           3          4    628.0         620   
3           4  2008      1           3          4   1829.0        1755   
4           5  2008      1           3          4   1940.0        1915   
5           6  2008      1           3          4   1937.0        1830   
6          10  2008      1           3          4    706.0         700   
7          11  2008      1           3          4   1644.0        1510   
8          15  2008      1           3          4   1029.0        1020   
9          16  2008      1           3          4   1452.0        1425   

   ArrTime  CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  \
0   2211.0        2225            WN        335  N712SW              128.0   
1   1002.0        1000            WN       3231  N772SW              128.0   
2    804.0         750            WN        448  N428WN               96.0   
3   1959.0        1925            WN       3920  N464WN               90.0   
4   2121.0        2110            WN        378  N726SW              101.0   
5   2037.0        1940            WN        509  N763SW              240.0   
6    916.0         915            WN        100  N690SW              130.0   
7   1845.0        1725            WN       1333  N334SW              121.0   
8   1021.0        1010            WN       2272  N263WN               52.0   
9   1640.0        1625            WN        675  N286WN              228.0   

   CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  \
0           150.0    116.0     -14.0       8.0    IAD  TPA       810     4.0   
1           145.0    113.0       2.0      19.0    IAD  TPA       810     5.0   
2            90.0     76.0      14.0       8.0    IND  BWI       515     3.0   
3            90.0     77.0      34.0      34.0    IND  BWI       515     3.0   
4           115.0     87.0      11.0      25.0    IND  JAX       688     4.0   
5           250.0    230.0      57.0      67.0    IND  LAS      1591     3.0   
6           135.0    106.0       1.0       6.0    IND  MCO       828     5.0   
7           135.0    107.0      80.0      94.0    IND  MCO       828     6.0   
8            50.0     37.0      11.0       9.0    IND  MDW       162     6.0   
9           240.0    213.0      15.0      27.0    IND  PHX      1489     7.0   

   TaxiOut  Cancelled CancellationCode  Diverted  CarrierDelay  WeatherDelay  \
0      8.0          0                N         0           NaN           NaN   
1     10.0          0                N         0           NaN           NaN   
2     17.0          0                N         0           NaN           NaN   
3     10.0          0                N         0           2.0           0.0   
4     10.0          0                N         0           NaN           NaN   
5      7.0          0                N         0          10.0           0.0   
6     19.0          0                N         0           NaN           NaN   
7      8.0          0                N         0           8.0           0.0   
8      9.0          0                N         0           NaN           NaN   
9      8.0          0                N         0           3.0           0.0   

   NASDelay  SecurityDelay  LateAircraftDelay  
0       NaN            NaN                NaN  
1       NaN            NaN                NaN  
2       NaN            NaN                NaN  
3       0.0            0.0               32.0  
4       NaN            NaN                NaN  
5       0.0            0.0               47.0  
6       NaN            NaN                NaN  
7       0.0            0.0               72.0  
8       NaN            NaN                NaN  
9       0.0            0.0               12.0

In [10]:
retrasos_full.CancellationCode.unique()   # Valores únicos en el campo "Cancellation"

array(['N', 'A', 'B', 'C'], dtype=object)

In [11]:
retrasos_full.info()  # Miramos el tipo de dato de cada columna.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1936758 entries, 0 to 1936757
Data columns (total 30 columns):
 #   Column             Dtype  
---  ------             -----  
 0   Unnamed: 0         int64  
 1   Year               int64  
 2   Month              int64  
 3   DayofMonth         int64  
 4   DayOfWeek          int64  
 5   DepTime            float64
 6   CRSDepTime         int64  
 7   ArrTime            float64
 8   CRSArrTime         int64  
 9   UniqueCarrier      object 
 10  FlightNum          int64  
 11  TailNum            object 
 12  ActualElapsedTime  float64
 13  CRSElapsedTime     float64
 14  AirTime            float64
 15  ArrDelay           float64
 16  DepDelay           float64
 17  Origin             object 
 18  Dest               object 
 19  Distance           int64  
 20  TaxiIn             float64
 21  TaxiOut            float64
 22  Cancelled          int64  
 23  CancellationCode   object 
 24  Diverted           int64  
 25  CarrierDelay      

Observamos los siguientes campos y/o tipos de datos:

+ El número de registro o número de entrada.
+ Fechas de vuelo (pero no en formato fecha, sino desglosadas en diferentes campos como enteros que corresponden al año, mes, día del mes y día de la semana).
+ Horas previstas y reales de los vuelos (de salida y llegada). Están indicadas como "float" o "int", no como hora.
+ Identificadores como el de de la compañía (object/"string"), numero de vuelo, número de Id de avión (i.e. el aparato propiamente dicho).
+ Duraciones previstas y reales de los vuelos (como float).
+ El tiempo que el avión realmente está en el aire durante el viaje (float).
+ El retraso en la salida, y en la llegada (float).
+ Origen, destinación (como object/"string") y distancia recorrida (número entero).
+ Tiempo de circulación sobre tierra (modo "taxi"), tanto a la salida como a la llegada (como "float").
+ Cancelación o no del vuelo (como entero "uno o cero"), y un código que puede ser, "N", "A", "B" o "C" (valores obtenidos ejecutando: "retrasos_full.CancellationCode.unique()")
+ Si se desvió el vuelo o no (como entero "uno o cero").
+ Desglosado por minutos, si se conocen, las causas del retraso: la compañía, el tiempo, el "NAS" (National Airspace System -entiendo que serán motivos regulatorios-), motivos de seguridad, o por retraso de avión (entiendo que se refiere a la llegada del avión desde su destino anterior). Valor dado como "float".


A continuación procedamos a obtener un dataframe reducido. Para eso haremos varias cosas:
+ Eliminaremos las filas que contengan registros vacíos ("NaN").
+ Pero ojo, antes, como los strings vacíos no se reconocen como tal, hemos de "traducirlos" a NaN.
+ Luego hacemos, ya sí, la eliminación de filas.
+ Tomamos la muestra de 200000 registros completos (ya ninguno tendrá NaN)
+ Eliminamos columnas redundantes o a las que no podamos dar utilidad.


Tenemos cinco campos que son strings ("object", en nomenclatura Pandas), son los siguientes:
"UniqueCarrier", "TailNum", "Origin", "Dest", "CancellationCode".
Hacemos la substitución de los "strings" vacíos en cada una de esas columnas por "NaN". 

In [12]:
len(retrasos_full)

1936758

In [13]:
campos_strings = "UniqueCarrier", "TailNum", "Origin", "Dest", "CancellationCode"

for element in campos_strings:  # Sustituimos strings vacíos por NaN
    retrasos_full[element].replace('', np.nan, inplace=True)

In [14]:
retrasos_full.dropna(axis=0, inplace=True)  # Eliminamos filas que contienen NaN.

Adicionalmente, podemos eliminar filas duplicadas, si es que las hubiera.

In [15]:
retrasos_full.drop_duplicates(inplace=True)

In [16]:
len(retrasos_full)  # Podemos comprobar como se han eliminado 689272 registros, aprox. un 35% de la muestra original.

1247486

Eliminaremos las siguientes columnas:
    
+ "Unnamed: 0", porque es redundante, no nos aporta información, no se sabe muy bien qué es ese número de orden (algún identificador).
+ "Year", porque todos los vuelos son del 2008, no aporta información diferenciadora.
+ "DepTime" y "CRSDepTime" (hora real y prevista de salida), porque servirían para para obtener "DepDelay", pero como ya tenemos este dato, son redundantes.
+ "ArrTime" y "CRSArrTime" (hora real y prevista de llegada) son redudantes por el mismo motivo, si nos damos cuenta de que ya tenemos el campo "ArrDelay".
+ "FlightNum" y "TailNum" solo son identificadores concretos que interpreto que no nos aportarán información sobre las tendencias generales. En cambio, sí nos quedamos con la compañía, por si vemos un patrón (p.ej. compañía con más retrasos que otras).

Pendiente de si al final usamos todas las demás o no, estas pienso que son las más descartables.

In [17]:
retrasos_reducido = retrasos_full.sample(200000)  # Creamos la muestra reducida que pedía el enunciado.
# Nota: después de consultarlo, es que hice el muestreo DESPUÉS de eliminar las filas que contenían campos vacíos.

In [18]:
# Eliminamos las columnas especificando sus nombres y "axis = 1":
retrasos_reducido = retrasos_reducido.drop(['Unnamed: 0', 'Year', 'DepTime', 'CRSDepTime', 'ArrTime', 'CRSArrTime', 'FlightNum', 'TailNum'], axis=1)

In [19]:
len(retrasos_reducido)

200000

In [20]:
retrasos_reducido

Month  DayofMonth  DayOfWeek UniqueCarrier  ActualElapsedTime  \
583514       4          11          5            WN              121.0   
1749467     12          18          4            WN               48.0   
1724701     11          30          7            CO              230.0   
1196277      7          29          2            FL              110.0   
702792       4           9          3            AA              264.0   
...        ...         ...        ...           ...                ...   
1873268     12          15          1            MQ               40.0   
999984       6           8          7            FL               79.0   
68220        1          18          5            UA              280.0   
977457       6          15          7            US               77.0   
1776449     12          21          7            XE              118.0   

         CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  \
583514            120.0    105.0      22.0      21.0    BWI  MDW       611   
1749467            60.0     33.0      55.0      67.0    ONT  LAS       197   
1724701           200.0    168.0     157.0     127.0    DFW  EWR      1372   
1196277            93.0     78.0      24.0       7.0    TPA  ATL       406   
702792            250.0    247.0      52.0      38.0    STL  SEA      1710   
...                 ...      ...       ...       ...    ...  ...       ...   
1873268            50.0     20.0      15.0      25.0    ORD  MKE        67   
999984             65.0     51.0     149.0     135.0    ATL  CLT       227   
68220             283.0    246.0      50.0      53.0    ORD  OAK      1835   
977457             73.0     43.0      73.0      69.0    LAS  PHX       256   
1776449           137.0     99.0      71.0      90.0    COS  IAH       809   

         TaxiIn  TaxiOut  Cancelled CancellationCode  Diverted  CarrierDelay  \
583514      4.0     12.0          0                N         0          16.0   
1749467     7.0      8.0          0                N         0          11.0   
1724701     7.0     55.0          0                N         0           0.0   
1196277     8.0     24.0          0                N         0           7.0   
702792      6.0     11.0          0                N         0           5.0   
...         ...      ...        ...              ...       ...           ...   
1873268     5.0     15.0          0                N         0           0.0   
999984      5.0     23.0          0                N         0           0.0   
68220      12.0     22.0          0                N         0          50.0   
977457     10.0     24.0          0                N         0           1.0   
1776449    10.0      9.0          0                N         0           0.0   

         WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay  
583514            0.0       1.0            0.0                5.0  
1749467           0.0       0.0            0.0               44.0  
1724701           0.0     157.0            0.0                0.0  
1196277           0.0      17.0            0.0                0.0  
702792            0.0      14.0            0.0               33.0  
...               ...       ...            ...                ...  
1873268           2.0       0.0            0.0               13.0  
999984            0.0      14.0            0.0              135.0  
68220             0.0       0.0            0.0                0.0  
977457            0.0       4.0            0.0               68.0  
1776449           0.0      71.0            0.0                0.0  

[200000 rows x 22 columns]

Liberamos de la memoria el dataframe original, más grande:

In [21]:
del retrasos_full
gc.collect()
retrasos_full=pd.DataFrame()
# Con esto pandas ya no guarda en su memoria el dataframe completo.

In [22]:
retrasos_full # Vemos que el dataframe original ya está vacío.

Empty DataFrame
Columns: []
Index: []

## Ejercicio 2:

Hacer un informe completo del dataset:
    
+ Resumir estadísticamente el dataset y las columnas de interés. Hacer un análisis estadístico de lo que sea relevante.
+ Hallar datos que faltan en cada columna.  [No posible aquí: se me dijo hacer la muestra de 200000 con registros completos.]
+ Crear columnas nuevas a partir de las iniciales (velocidad media, retrasos, otros, etc.).
+ Tabular aerolíneas con más retrasos.
+ ¿Qué vuelos son los más largos? ¿Y los más retrasados? Buscar cuáles son estas rutas.
+ Asimismo, añadir los comentarios que sean relevantes.




Echemos primero un vistazo al dataset reducido, desde el punto de vista de los estadísticos generales: medias, desviación estándar, cuartiles, mediana, etc.

In [23]:
retrasos_reducido.describe()

Month     DayofMonth      DayOfWeek  ActualElapsedTime  \
count  200000.000000  200000.000000  200000.000000      200000.000000   
mean        6.074980      15.702195       3.974045         135.214040   
std         3.515598       8.782858       1.992529          72.114619   
min         1.000000       1.000000       1.000000          16.000000   
25%         3.000000       8.000000       2.000000          83.000000   
50%         6.000000      16.000000       4.000000         118.000000   
75%         9.000000      23.000000       6.000000         167.000000   
max        12.000000      31.000000       7.000000        1114.000000   

       CRSElapsedTime        AirTime      ArrDelay       DepDelay  \
count   200000.000000  200000.000000  200000.00000  200000.000000   
mean       131.579295     107.234985      63.39679      59.762045   
std         69.315257      68.003078      60.83598      59.861923   
min          9.000000       1.000000      15.00000       6.000000   
25%         80.000000      58.000000      26.00000      24.000000   
50%        115.000000      89.000000      43.00000      41.000000   
75%        161.000000     136.000000      79.00000      75.000000   
max        660.000000    1091.000000    1707.00000    1710.000000   

            Distance         TaxiIn        TaxiOut  Cancelled  Diverted  \
count  200000.000000  200000.000000  200000.000000   200000.0  200000.0   
mean      739.995675       7.284330      20.694725        0.0       0.0   
std       558.313704       6.009174      16.684745        0.0       0.0   
min        11.000000       0.000000       0.000000        0.0       0.0   
25%       334.000000       4.000000      11.000000        0.0       0.0   
50%       594.000000       6.000000      16.000000        0.0       0.0   
75%       967.000000       8.000000      24.000000        0.0       0.0   
max      4962.000000     240.000000     371.000000        0.0       0.0   

        CarrierDelay   WeatherDelay       NASDelay  SecurityDelay  \
count  200000.000000  200000.000000  200000.000000  200000.000000   
mean       19.201865       3.709380      15.062590       0.093410   
std        43.594423      21.245429      33.959314       2.114412   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         2.000000       0.000000       2.000000       0.000000   
75%        21.000000       0.000000      15.000000       0.000000   
max      1707.000000    1352.000000     865.000000     284.000000   

       LateAircraftDelay  
count      200000.000000  
mean           25.329545  
std            42.254734  
min             0.000000  
25%             0.000000  
50%             8.000000  
75%            33.000000  
max          1303.000000

Hay una serie de campos para los cuales el resumen obtenido no nos aporta información útil (los referidos al número de mes, de día del mes, de día de la semana... ni tampoco los referidos a valores "binarios" de "sí o no" o "1 o 0", ni los referidos a códigos o identificadores).

Si nos fijamos en los valores numéricos referidos a retrasos y tiempos (p.ej. tiempo de vuelo, de taxi, minutos de retraso, etc.), sí podemos decir algunas cosas:
+ Podemos constatar que en los registros que tenemos, hay generalmente un retraso importante, ya sea "ArrDelay", "DepDelay": el retraso medio es en ambos casos cercano a los 60 minutos.
+ La aportación promedio más alta a esos retrasos es de la compañía (CarrierDelay, unos 19 minutos), la agencia reguladora "NAS" ("NASDelay", unos 15 minutos) y de la llegada tardía del avíon que ha de transportar al pasaje ("LateAircraftDelay", unos 25 min.).
+ Observando los valores máximos y mínimos podemos que el retraso máximo tanto de llegada ("ArrDelay") como de salida ("DepDelay") es cercano a los 1500 minutos.
+ Además, los retrasos mínimos son de 15 y 6 minutos, respectivamente. Y a partir de ahí, más. Esto implica que CADA APARICIÓN en esta lista conlleva un retraso.
+ Vemos que el vuelo más largo en cuanto a kilometraje cubre 4962 kilómetros, y en cuanto a tiempo total real de viaje (dejando de lado retraso a la llegada y salida), el más largo en cuanto a tiempo dura 727 minutos.
+ El vuelo más corto en cuanto a tiempo dura 14 minutos.
+ La distancia más corta recorrida es de 31 km.
+ Nos podemos fijar también en el valor de la mediana (marcado como "50%"), que nos dice el valor que hay en medio de cada distribución. Un patrón que se repite es que nuestros valores medianos son más bajos que los valores medios. Esto significa que hay algunos "outliers" que están empujando la media hacia arriba.
+ Podemos constatar una importante dispersión de los valores, como indican los elevados valores de desviación estándar. Por ejemplo, en el tiempo total de los vuelos ("ActualElapsedTime") tenemos una desviación estándar de unos 72 minutos, y eso teniendo esa variable una media de 135 minutos: es una desviación muy relevante, en comparación. Y este patrón se repite con más o menos intensidad para las otras variables numéricas, que también tienen mucha dispersión. En el caso de las últimas cinco columnas, esto es todavía mayor.
+ De hecho, si miramos las últimas columnas, podemos darnos cuenta que en promedio los factores que aportan más retraso son "CarrierDelay", "NASDelay", y "LateAircraftDelay". Igualmente, si nos fijamos en la mediana, podemos inferir que en al menos la mitad de ocasiones, el impacto de estos factores es relativamente bajo (salvo "LateAircraftDelay", que imputa unos ocho minutos en el caso mediano).
+ Fijándonos en "TaxiIn" y "TaxiOut" deducimos que el tiempo de viaje sobre ruedas es mucho menor a la llegadas de las vuelos que a su salida.


Como parte del ejercicio, creemos algunas nuevas columnas calculadas a partir de las iniciales. Podemos calcular, por ejemplo: la velocidad media de un vuelo, o la diferencia entre la duración real de un vuelo y la prevista (que viene a ser el tiempo recuperado o perdido desde la salida hasta la llegada.)
Podemos llamarlas:
v_media, dur_real_vs_prevista

In [24]:
retrasos_reducido["v_media"] = retrasos_reducido["Distance"]/(retrasos_reducido["ActualElapsedTime"]/60)
retrasos_reducido["dur_real_vs_prevista"] = retrasos_reducido["ActualElapsedTime"] - retrasos_reducido["CRSElapsedTime"]


In [25]:
retrasos_reducido.head()

Month  DayofMonth  DayOfWeek UniqueCarrier  ActualElapsedTime  \
583514       4          11          5            WN              121.0   
1749467     12          18          4            WN               48.0   
1724701     11          30          7            CO              230.0   
1196277      7          29          2            FL              110.0   
702792       4           9          3            AA              264.0   

         CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  \
583514            120.0    105.0      22.0      21.0    BWI  MDW       611   
1749467            60.0     33.0      55.0      67.0    ONT  LAS       197   
1724701           200.0    168.0     157.0     127.0    DFW  EWR      1372   
1196277            93.0     78.0      24.0       7.0    TPA  ATL       406   
702792            250.0    247.0      52.0      38.0    STL  SEA      1710   

         TaxiIn  TaxiOut  Cancelled CancellationCode  Diverted  CarrierDelay  \
583514      4.0     12.0          0                N         0          16.0   
1749467     7.0      8.0          0                N         0          11.0   
1724701     7.0     55.0          0                N         0           0.0   
1196277     8.0     24.0          0                N         0           7.0   
702792      6.0     11.0          0                N         0           5.0   

         WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay     v_media  \
583514            0.0       1.0            0.0                5.0  302.975207   
1749467           0.0       0.0            0.0               44.0  246.250000   
1724701           0.0     157.0            0.0                0.0  357.913043   
1196277           0.0      17.0            0.0                0.0  221.454545   
702792            0.0      14.0            0.0               33.0  388.636364   

         dur_real_vs_prevista  
583514                    1.0  
1749467                 -12.0  
1724701                  30.0  
1196277                  17.0  
702792                   14.0

Se han incorporado al final del dataframe dos columnas más. Saquemos algunas conclusiones adicionales. Por ejemplo, a partir de la primera de las nuevas columnas que hemos creado, ¿cuál es la velocidad media más alta?

In [26]:
retrasos_reducido['v_media'].max()  # Nos da la velocidad media más alta (en millas por hora).

980.0

In [27]:
retrasos_reducido.loc[retrasos_reducido['v_media'] == retrasos_reducido['v_media'].max()]

Month  DayofMonth  DayOfWeek UniqueCarrier  ActualElapsedTime  \
331502      2           3          7            9E               30.0   

        CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  \
331502           100.0      2.0      60.0     130.0    MEM  DSM       490   

        TaxiIn  TaxiOut  Cancelled CancellationCode  Diverted  CarrierDelay  \
331502     5.0     23.0          0                N         0          60.0   

        WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay  v_media  \
331502           0.0       0.0            0.0                0.0    980.0   

        dur_real_vs_prevista  
331502                 -70.0

La velocidad media más alta corresponde al vuelo entre el origen MEM y el destino DSM, con una distancia total de 490 km (nº de "row label" 331502).

A partir de la otra columna que hemos creado, ¿cuál es el vuelo en el que se ha recuperado más tiempo durante el vuelo?

In [76]:
retrasos_reducido.loc[retrasos_reducido['dur_real_vs_prevista'] == retrasos_reducido['dur_real_vs_prevista'].min()]

Month  DayofMonth  DayOfWeek UniqueCarrier  ActualElapsedTime  \
522132      3           9          7            9E              148.0   

        CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  \
522132           290.0     90.0     101.0     243.0    PIT  ATL       526   

        TaxiIn  TaxiOut  Cancelled CancellationCode  Diverted  CarrierDelay  \
522132    32.0     26.0          0                N         0           0.0   

        WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay     v_media  \
522132           0.0       0.0            0.0              101.0  213.243243   

        dur_real_vs_prevista  
522132                -142.0

El vuelo en el que se ha recuperado más tiempo durante el trayecto (comparando con el tiempo previsto de viaje) es el vuelo con "row id" 522132, entre Pittsburgh y Atlanta.

Vuelo más largo:

In [28]:
retrasos_reducido['Distance'].max()  # Distancia más larga, en millas.

4962

In [83]:
retrasos_reducido.loc[retrasos_reducido['Distance'] == retrasos_reducido['Distance'].max()]

Month  DayofMonth  DayOfWeek UniqueCarrier  ActualElapsedTime  \
1621350     10          16          4            CO              646.0   
1422029      8          20          3            CO              631.0   
1930578     12           8          1            CO              615.0   
721208       4          10          4            CO              631.0   
362529       2          21          4            CO              697.0   
1930729     12           7          7            CO              637.0   
719889       4          20          7            CO              594.0   
716859       4          13          7            CO              633.0   
716110       4          26          6            CO              644.0   
1070004      6           5          4            CO              658.0   
1725876     11          22          6            CO              645.0   
1726009     11          21          5            CO              661.0   
1418032      8           2          6            CO              750.0   
875732       5           3          6            CO              568.0   
1075813      6          14          6            CO              651.0   
1251924      7          16          3            CO              627.0   
1066708      6          19          4            CO              666.0   
556384       3          25          2            CO              659.0   
556358       3          18          2            CO              559.0   
1415175      8           7          4            CO              626.0   
1518539      9          11          4            CO              626.0   
1413415      8           3          7            CO              660.0   
174818       1          22          2            CO              635.0   

         CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  \
1621350           635.0    600.0      26.0      15.0    EWR  HNL      4962   
1422029           635.0    569.0      44.0      48.0    EWR  HNL      4962   
1930578           565.0    583.0     299.0     249.0    HNL  EWR      4962   
721208            635.0    605.0      27.0      31.0    EWR  HNL      4962   
362529            660.0    664.0      54.0      17.0    EWR  HNL      4962   
1930729           660.0    610.0      37.0      60.0    EWR  HNL      4962   
719889            575.0    554.0     220.0     201.0    HNL  EWR      4962   
716859            635.0    605.0      25.0      27.0    EWR  HNL      4962   
716110            635.0    605.0      19.0      10.0    EWR  HNL      4962   
1070004           635.0    606.0     123.0     100.0    EWR  HNL      4962   
1725876           660.0    609.0      23.0      38.0    EWR  HNL      4962   
1726009           660.0    622.0      38.0      37.0    EWR  HNL      4962   
1418032           635.0    597.0     388.0     273.0    EWR  HNL      4962   
875732            575.0    545.0      16.0      23.0    HNL  EWR      4962   
1075813           635.0    617.0      26.0      10.0    EWR  HNL      4962   
1251924           635.0    592.0      15.0      23.0    EWR  HNL      4962   
1066708           635.0    620.0     221.0     190.0    EWR  HNL      4962   
556384            660.0    632.0      61.0      62.0    EWR  HNL      4962   
556358            565.0    522.0      16.0      22.0    HNL  EWR      4962   
1415175           635.0    598.0      26.0      35.0    EWR  HNL      4962   
1518539           635.0    603.0      27.0      36.0    EWR  HNL      4962   
1413415           635.0    629.0      34.0       9.0    EWR  HNL      4962   
174818            660.0    607.0      56.0      81.0    EWR  HNL      4962   

         TaxiIn  TaxiOut  Cancelled CancellationCode  Diverted  CarrierDelay  \
1621350     6.0     40.0          0                N         0           0.0   
1422029     6.0     56.0          0                N         0          44.0   
1930578    13.0     19.0          0                N         0         249.0   
721208      8.0     18.0          0    

Vemos que la distancia más larga son 4962 km, pero corresponde a múltiples vuelos entre dos destinos: EWR (New York Newark) y HNL (Honolulu), y claaro, la ruta inversa. Si quisiéramos obtener el "row id" de cada uno, podríamos hacerlo así:

In [30]:
row_num = retrasos_reducido[retrasos_reducido['Distance'] == retrasos_reducido['Distance'].max()].index 

In [82]:
row_num

Index([1621350, 1422029, 1930578,  721208,  362529, 1930729,  719889,  716859,
        716110, 1070004, 1725876, 1726009, 1418032,  875732, 1075813, 1251924,
       1066708,  556384,  556358, 1415175, 1518539, 1413415,  174818],
      dtype='int64')

¿Cuál es el vuelo con más retraso, cuánto retraso lleva? El vuelo con más retraso implica que es el que llega con un mayor "ArrDelay". Es decir:

In [81]:
retrasos_reducido['ArrDelay'].max()  # Valor en minutos.

1707.0

In [80]:
# Si recuperamos todo el registro:
retrasos_reducido.loc[retrasos_reducido['ArrDelay'] == retrasos_reducido['ArrDelay'].max()]

Month  DayofMonth  DayOfWeek UniqueCarrier  ActualElapsedTime  \
1009553      6          20          5            MQ               72.0   

         CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  \
1009553            75.0     50.0    1707.0    1710.0    LIT  DFW       304   

         TaxiIn  TaxiOut  Cancelled CancellationCode  Diverted  CarrierDelay  \
1009553     6.0     16.0          0                N         0        1707.0   

         WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay     v_media  \
1009553           0.0       0.0            0.0                0.0  253.333333   

         dur_real_vs_prevista  
1009553                  -3.0

Vemos el vuelo de nuestra muestra de 200k registros que lleva más retraso corresponde al vuelo con "Row Index" 1009553, entre HNL y MSP.

In [79]:
retrasos_reducido[retrasos_reducido['ArrDelay'] == retrasos_reducido['ArrDelay'].max()].index 

Index([1009553], dtype='int64')

Hacer una tabla de las compañías con más retraso (interpretado como retraso promedio por vuelo):

In [111]:
retrasos_reducido.groupby(['UniqueCarrier'])['ArrDelay'].mean()

UniqueCarrier
9E    66.891570
AA    65.968220
AQ    47.982456
AS    56.902739
B6    78.949033
CO    68.127727
DL    60.335610
EV    66.221119
F9    44.927508
FL    64.925208
HA    55.475078
MQ    63.179922
NW    60.709667
OH    66.540493
OO    66.105530
UA    69.410331
US    59.347167
WN    52.608008
XE    70.009209
YV    70.139441
Name: ArrDelay, dtype: float64

Podemos concluir que la compañía con más retraso promedio es "9E".

Y, ¿cuáles son las rutas más largas? ¿Y cuáles las rutas que acumulan más retrasos?".

In [48]:
retrasos_reducido.loc[retrasos_reducido['Distance'] == retrasos_reducido['Distance'].max()]

Month  DayofMonth  DayOfWeek UniqueCarrier  ActualElapsedTime  \
1621350     10          16          4            CO              646.0   
1422029      8          20          3            CO              631.0   
1930578     12           8          1            CO              615.0   
721208       4          10          4            CO              631.0   
362529       2          21          4            CO              697.0   
1930729     12           7          7            CO              637.0   
719889       4          20          7            CO              594.0   
716859       4          13          7            CO              633.0   
716110       4          26          6            CO              644.0   
1070004      6           5          4            CO              658.0   
1725876     11          22          6            CO              645.0   
1726009     11          21          5            CO              661.0   
1418032      8           2          6            CO              750.0   
875732       5           3          6            CO              568.0   
1075813      6          14          6            CO              651.0   
1251924      7          16          3            CO              627.0   
1066708      6          19          4            CO              666.0   
556384       3          25          2            CO              659.0   
556358       3          18          2            CO              559.0   
1415175      8           7          4            CO              626.0   
1518539      9          11          4            CO              626.0   
1413415      8           3          7            CO              660.0   
174818       1          22          2            CO              635.0   

         CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  \
1621350           635.0    600.0      26.0      15.0    EWR  HNL      4962   
1422029           635.0    569.0      44.0      48.0    EWR  HNL      4962   
1930578           565.0    583.0     299.0     249.0    HNL  EWR      4962   
721208            635.0    605.0      27.0      31.0    EWR  HNL      4962   
362529            660.0    664.0      54.0      17.0    EWR  HNL      4962   
1930729           660.0    610.0      37.0      60.0    EWR  HNL      4962   
719889            575.0    554.0     220.0     201.0    HNL  EWR      4962   
716859            635.0    605.0      25.0      27.0    EWR  HNL      4962   
716110            635.0    605.0      19.0      10.0    EWR  HNL      4962   
1070004           635.0    606.0     123.0     100.0    EWR  HNL      4962   
1725876           660.0    609.0      23.0      38.0    EWR  HNL      4962   
1726009           660.0    622.0      38.0      37.0    EWR  HNL      4962   
1418032           635.0    597.0     388.0     273.0    EWR  HNL      4962   
875732            575.0    545.0      16.0      23.0    HNL  EWR      4962   
1075813           635.0    617.0      26.0      10.0    EWR  HNL      4962   
1251924           635.0    592.0      15.0      23.0    EWR  HNL      4962   
1066708           635.0    620.0     221.0     190.0    EWR  HNL      4962   
556384            660.0    632.0      61.0      62.0    EWR  HNL      4962   
556358            565.0    522.0      16.0      22.0    HNL  EWR      4962   
1415175           635.0    598.0      26.0      35.0    EWR  HNL      4962   
1518539           635.0    603.0      27.0      36.0    EWR  HNL      4962   
1413415           635.0    629.0      34.0       9.0    EWR  HNL      4962   
174818            660.0    607.0      56.0      81.0    EWR  HNL      4962   

         TaxiIn  TaxiOut  Cancelled CancellationCode  Diverted  CarrierDelay  \
1621350     6.0     40.0          0                N         0           0.0   
1422029     6.0     56.0          0                N         0          44.0   
1930578    13.0     19.0          0                N         0         249.0   
721208      8.0     18.0          0    

In [49]:
# Podemos ver los valores más altos de forma no redundante, por ejemplo así. 
sorted(retrasos_reducido['Distance'].unique(), reverse=True)

[4962,
 4502,
 4243,
 4213,
 4184,
 3972,
 3904,
 3784,
 3711,
 3417,
 3414,
 3386,
 3365,
 3329,
 3303,
 3266,
 3110,
 3043,
 2994,
 2986,
 2979,
 2936,
 2917,
 2860,
 2846,
 2845,
 2797,
 2777,
 2762,
 2724,
 2704,
 2701,
 2693,
 2689,
 2688,
 2677,
 2640,
 2615,
 2614,
 2611,
 2603,
 2602,
 2588,
 2586,
 2585,
 2583,
 2576,
 2569,
 2565,
 2562,
 2556,
 2553,
 2551,
 2548,
 2541,
 2537,
 2534,
 2530,
 2527,
 2521,
 2518,
 2504,
 2496,
 2486,
 2475,
 2466,
 2465,
 2462,
 2458,
 2457,
 2454,
 2447,
 2446,
 2445,
 2433,
 2430,
 2425,
 2421,
 2419,
 2416,
 2409,
 2408,
 2406,
 2404,
 2401,
 2400,
 2398,
 2381,
 2378,
 2369,
 2367,
 2363,
 2358,
 2356,
 2349,
 2345,
 2342,
 2338,
 2335,
 2329,
 2327,
 2311,
 2306,
 2304,
 2300,
 2298,
 2296,
 2295,
 2288,
 2283,
 2282,
 2279,
 2277,
 2254,
 2253,
 2248,
 2244,
 2239,
 2237,
 2227,
 2217,
 2213,
 2184,
 2182,
 2176,
 2175,
 2174,
 2172,
 2165,
 2161,
 2158,
 2155,
 2153,
 2148,
 2139,
 2136,
 2133,
 2130,
 2125,
 2116,
 2106,
 2105,
 2092,

In [141]:
# Lo siguiente nos muestra los vuelos concretos que cubren las tres distancias más largas que hemos visto hace un momento:
for element in [4962, 4502, 4243]:
    print(retrasos_reducido[["Origin", "Dest"]].loc[retrasos_reducido['Distance'] == element])

        Origin Dest
1621350    EWR  HNL
1422029    EWR  HNL
1930578    HNL  EWR
721208     EWR  HNL
362529     EWR  HNL
1930729    EWR  HNL
719889     HNL  EWR
716859     EWR  HNL
716110     EWR  HNL
1070004    EWR  HNL
1725876    EWR  HNL
1726009    EWR  HNL
1418032    EWR  HNL
875732     HNL  EWR
1075813    EWR  HNL
1251924    EWR  HNL
1066708    EWR  HNL
556384     EWR  HNL
556358     HNL  EWR
1415175    EWR  HNL
1518539    EWR  HNL
1413415    EWR  HNL
174818     EWR  HNL
        Origin Dest
725194     HNL  ATL
1426988    ATL  HNL
572709     HNL  ATL
723883     ATL  HNL
370960     ATL  HNL
1262781    ATL  HNL
1731008    ATL  HNL
97783      ATL  HNL
651052     ATL  HNL
1264888    HNL  ATL
723215     HNL  ATL
99324      ATL  HNL
808779     HNL  ATL
1577460    HNL  ATL
652235     ATL  HNL
182317     ATL  HNL
881738     ATL  HNL
881188     HNL  ATL
986846     ATL  HNL
988603     ATL  HNL
986469     HNL  ATL
1626499    ATL  HNL
1350424    ATL  HNL
181464     ATL  HNL
1841578    ATL  HNL


De lo anterior, podemos deducir que los vuelos más largos son los que van entre New York Newark y Honolulu (4962), entre Hartsfield-Jackson Atlanta y Honolulu (4502), y entre Chicago O'Hare y Honolulu (4243), y obviamente, las rutas inversas.

Para ver qué RUTAS acumulan más número de retrasos, hemos de sumar el retraso de las rutas. En este caso consideraremos como una ruta aquella que transcurre en la misma dirección y sentido (i.e. Honolulu -> Newark diremos que es una ruta "diferente" que la inversa).

In [135]:
# Agrupamos por Origen y Destino.
# Esto da los retrasos acumulados de cada ruta concreta.
retrasos_reducido.groupby(['Origin', 'Dest'])['ArrDelay'].sum()

Origin  Dest
ABE     ATL     1342.0
        CLE      765.0
        CLT      322.0
        CVG       25.0
        DTW      950.0
                 ...  
YUM     IPL       93.0
        LAS       20.0
        LAX     1123.0
        PHX     1138.0
        SLC      101.0
Name: ArrDelay, Length: 4796, dtype: float64

In [138]:
retrasos_reducido.groupby(['Origin', 'Dest'])['ArrDelay'].sum().sort_values(ascending=False)

Origin  Dest
LGA     ORD     39338.0
ORD     LGA     38031.0
LAX     SFO     36304.0
ORD     EWR     34487.0
SFO     LAX     32907.0
                 ...   
SLC     FAR        15.0
MEM     BDL        15.0
CLE     SJU        15.0
RKS     GCC        15.0
MCO     SAN        15.0
Name: ArrDelay, Length: 4796, dtype: float64

La ruta que acumula más minutos de retraso, por lo tanto, es entre New York La Guardia (LGA) y Chicago O'Hare.    

## Ejercicio 3:

Exportar el dataset y las nuevas columnas a Excel.

In [74]:
retrasos_reducido.to_excel("dataset_200k_retrasos.xlsx", sheet_name='retrasos_muestra_200k')